### PR TITLE
Updated Voby to 0.2.0

### PR DESCRIPTION
**Pull request by fabiospampinato**
I implemented a few optimizations here and there, maybe they'll make a measurable difference in the benchmark 🤔

### CODE REVIEW

**Comment by fabiospampinato** [comment]
v0.3.0 has a pretty amazing optimization, I've very excited to see that in the benchmark!

**Comment by fabiospampinato** [comment]
From running the benchmark locally it looks like the benchmark doesn't quite measure this area much, so the optimization doesn't make much of a difference in the benchmark 😢 It'd be great if a "replace all rows" test but with 10k rows were added, it seems to be missing currently.

**Comment by fabiospampinato** [comment]
Shipped another batch of tweaks and optimizations.

@krausest if possible it'd be awesome to get this merged before the results table is refreshed for Chrome 100 🙏

**Comment by fabiospampinato** [comment]
Expanding on the issue of the benchmark being somewhat creation-biased:

- The "append" test is also mainly measuring creation at the end of the day, and that's much more expensive than the diffing.
- The "clear rows" test is kind of giving bad results if nodes are reused, because if they can be reused there's a bit more work to do to recycle them, while if you can just throw them away it would be faster, the thing is recycling nodes ends up giving you much better performance at runtime when it matters, but that's almost entirely not measured by this benchmark, except for the "replace 1000 rows" test.

Basically it'd be nice if the benchmark measured more things similar to runtime performance, like complex edits that happen without refreshing the browser in between.

**Comment by fabiospampinato** [comment]
I'm getting some promising results locally with the last build. There were a handful of missing optimizations and some useless work that the previous builds were doing.

**Comment by krausest** [comment]
Screenshot from 2022-03-13 12-55-12
A nice improvement!

**Comment by krausest** [comment]
There one catch though. I think voby keeps the selected state on the row and not on the table. Thus I had to add note #800 .
Could you try to store the selected row in the model and then set the danger-class for the row that's equal to the selected row in the row render loop?

**Comment by fabiospampinato** [comment]
Nice! Thanks for merging. 

There's at least one more little optimization that would be worth implementing. Other than that one I'm running out of ideas though.

> Could you try to store the selected row in the model and then set the danger-class for the row that's equal to the selected row in the row render loop?

Sure, I can do that 👍

File: frameworks/keyed/voby/src/main.tsx [hunk 1]
```diff
@@ -100,25 +99,26 @@ const Model = (() => {
 
   const remove = ( id: string ): void => {
     const data = $data ();
-    const index = data.findIndex ( datum => datum.sample ().id === id );
-    $data ( [...data.slice ( 0, index ), ...data.slice ( index + 1 )] );
+    const index = data.findIndex ( datum => datum.id === id );
+    if ( index === -1 ) return;
+    $data ( data.slice ( 0, index ).concat ( data.slice ( index + 1 ) ) );
```
Comment (leeoniya) [inline frameworks/keyed/voby/src/main.tsx:104]: @fabiospampinato would the shorter `.splice()` be any slower?

Comment (fabiospampinato) [reply frameworks/keyed/voby/src/main.tsx:104]: @leeoniya I'm not sure, it might depend on where the deletion happens.

I went with slice/concat because I need to make a new array to have the observable detect it as a change so that it can notify all its subscribers. But now that I think about it it would probably be better to expose the method for notifying subscribers, so that I can just edit the array in place and notify them manually.

Comment (leeoniya) [reply frameworks/keyed/voby/src/main.tsx:104]: not sure the notification route would have any measurable effect in this specific case, but good to have as option in voby.

for this bench i'd probably leave it written as-is.

Comment (fabiospampinato) [reply frameworks/keyed/voby/src/main.tsx:104]: It might shave 1ms or so, but more importantly it would make the code cleaner.

Comment (leeoniya) [reply frameworks/keyed/voby/src/main.tsx:104]: > but more importantly it would make the code cleaner.

unless it's mutate & notify in some places, and shallow copy in other places :)

Comment (fabiospampinato) [reply frameworks/keyed/voby/src/main.tsx:104]: > unless it's mutate & notify in some places, and shallow copy in other places :)

A little messy, but cleaner than what the code looks like now. For the cleanest code when working with non-trivial data structures I think we ideally want to wrap the data structure in a proxy and extract observables out of it automatically, or something like that. I have another library for that but I haven't wired it with this one, it should be doable 🤔
